File: solutions/3106. Lexicographically Smallest String After Operations With Constraint/3106.cpp
// Time complexity: O(n)
// Space complexity: O(1)
class Solution {
 public:
  string getSmallestString(string s, int k) {
    for (char &c : s) {
      int l = c - 'a';
      int r = 'z' - c + 1;
      int minDistance = min(l, r);
      if (k >= minDistance) {
        k -= minDistance;
        c = 'a';
      } else {
        c = c - k;
        k = 0;
      }
    }

    return s;
  }
};

File: solutions/3106. Lexicographically Smallest String After Operations With Constraint/3106.py
# Time complexity: O(n)
# Space complexity: O(n)
class Solution:
  def getSmallestString(self, s: str, k: int) -> str:
    res = ""

    for c in s:
      minDistance = min(ord(c) - ord('a'), ord('z') - ord(c) + 1)

      if k >= minDistance:
        k -= minDistance
        res += 'a' if minDistance <= ord(c) - ord('a') else 'z'
      else:
        res += chr(ord(c) - k)
        k = 0

    return res